### DISCUSSION TITLE
Controlling multiple pins at once

### DISCUSSION THREAD

**Discussion by Tow96** [earlephilhower/arduino-pico#722] [General]
Hi Everyone,

In the regular arduino library, there is a way to read/write/set multiple pins at once by using the registers, for example using register B, you can read, write and setIO using: PINB PORTB and DDRB respectively.

Is there a way to do the same using the pi pico?

**Comment by Andy2No** [comment]
@Tow96 You can use the gpio functions, that are part of the SDK - you don't have to do anything special to use them:

https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__gpio.html

For example, if you want to output a byte on 8 consecutively numbered IO pins, you can use the  gpio_put_masked() function.  You give it a bit mask and a value, both 32 bit.

Here's what I did to drive an 8 bit R2R ladder DAC (a DAC made from resistors):

```
const int r2r_lowp = 0; // GPIO pins for R2R ladder DAC output
const int r2r_highp = 7;

uint32_t r2r_mask = 0;
int r2r_bits = r2r_highp - r2r_lowp + 1;
uint32_t r2r_range = (2 << r2r_bits) - 1;   // Max value for the number of bits

void r2r_out(uint16_t v) {  // allow for upto a 16 bit R2R ladder DAC (probably a practical limit)
  v= v << r2r_lowp;
  gpio_put_masked(r2r_mask, v);
}
// ...

void setup() {
  // Set output pins for R2R ladder DAC
  for (int p = r2r_lowp; p <= r2r_highp; p++) {
    r2r_mask |= (1 << p); // Make a mask for parallel I/O
    pinMode(p, OUTPUT);
  }
}

void loop() {
      int pot1 = analogRead(A1);
      r2r_out(pot1);  // Test the R2R DAC

      // You may want a short delay here, if not doing something else before / afterwards
}
```

The pin numbering is in gpio pin numbers, rather than pins numbered like on a chip.  It works as one 32 bit port, but not all pins are accessible, and you should mask off ones you're not using, when you write to them. 

You could use gpio functions to control the mode of the pins too, but pinMode() works fine.
